[v9.x backport] test: fix require-deps-deprecation for installed deps

Tiriel opened the pull request:
Test test-require-deps-deprecation.js was failing when user already had
node installed with acorn in require.resolve range.

Modified test to acknowledge the possibility and throw only if acorn is
found in the deps directory.

Also changed the deprecation test for v9.x: common.expectWarning was failing
because the required deps now throw ReferenceErrors when not properly called
internally in the right order.

PR-URL: https://github.com/nodejs/node/pull/17848
Fixes: https://github.com/nodejs/node/issues/17148
Reviewed-By: Tiancheng "Timothy" Gu <timothygu99@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

Original PR: https://github.com/nodejs/node/pull/17848

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
test

Tiriel commented:
I don't know if this is the correct way to adapt this test, but I figured it was the best.

For the explanation:
The initial commit could not be backported on v9.x . After some investigation, the problem did not come from the require test modification, but from the deprecation warning expectations. From what I saw, these expectations where failing because require is now throwing `ReferenceError` when the internal deps are not properly called (ie internally to Node, not in a loop, and in the proper order).
So, this solution appeared the best to me. But I'll gladly amend in any way, or even remove this PR if this cannot be backported as it is now.

lpinca commented on test/parallel/test-require-deps-deprecation.js:
```diff
@@ -0,0 +1,52 @@
+  }
+};
+
+assert.throws(throwingModules, ReferenceError);
```
This doesn't seem correct, it should be inside the the loop or it will only test the first `require()` call that throws.

Tiriel replied:
True, but that's one of the problem: I'm not sure every require throws.

From what I've seen, some tools throw (typically `v8/tools/profile`), some other don't. Should I remove every module not throwing? Or is there a better way to test this?

lpinca replied:
I'm not sure either, let's ping the reviewers of the original PR.
cc: @TimothyGu @cjihrig @Trott @jasnell

Trott replied:
The checking for `ReferenceError` appears to only check that `require(m)` is the wrong way to include two of the modules in the list. The correct way to include (for example) `'v8/tools/tickprocessor'` is not `require('v8/tools/tickprocessor')` but rather something like `eval(process.binding('natives')['v8/tools/tickprocessor'])` when run with `--expose-internals`.

I think the test for `ReferenceError` should be removed entirely. The important thing to test is that a DeprecationWarning is being emitted, using `common.expectsWarning()` or whatever it is that's in the `common` module for that. I'm going to change my review to Request Changes for that. Sorry that I missed that on first review!

Trott replied:
Oh, wait, I didn't approve it or even review this until now. It's only 8 hours old. Never mind. But still, the comment stands: Yeah, that should be changed. Sorry if I'm directly or indirectly responsible for that code in the first place.

lpinca replied:
@Trott this is an attempt to backport #17848.

Trott replied:
@lpinca Thanks. I don't see a `ReferenceError` check in this file on the master branch so now I'm even more confused. :-D

cjihrig replied:
I'm not sure this should actually be backported.

Tiriel commented:
@Trott That’s why I changed the test in fact: no warning is emitted, because an error is thrown as they are not properly required. 

The test was failing because no warning was emitted, so I removed the try-catch and got errors saying that CodeMap wasn’t defined in profile.js .

It seems this test is pointless if these deeps cannot be required without throwing. Or is there some other way?

Trott commented:
Ah, yeah, you'll need to remove any modules that aren't deprecated in this version of Node.js. If that means removing all of them, then yeah, we don't need the test in v9.x. :-D

Tiriel commented:
Made some more tests and I can confirm, the test on acorn still seems relevant, but I don't know if it's worth keeping it.

As for the other tests, no v8/tool seem to be emitting warnings on v9.x so I just dropped them.

Tiriel commented:
Just a small up.
Does anyone thinks this can/should be merged, or do I drop this PR?

Trott commented:
@nodejs/testing

BridgeAR commented:
Ping @nodejs/lts

jasnell commented:
Why is this targeting v9.x instead of master?
And if it is v9.x, then pinging @nodejs/lts is not necessary since 9.x is not an LTS branch.

richardlau commented:
> Why is this targeting v9.x instead of master?

This is a requested backport (https://github.com/nodejs/node/pull/17848#issuecomment-356451489) of a PR that has already landed on master. I've amended the PR title so that it is clearer.

MylesBorins commented:
Landed in b5752ee6a4

Tiriel commented:
Thanks!
